### PR TITLE
Process.run bugfixes and enhancements:

### PR DESCRIPTION
**Pull request by technorama**
Close the correct pipes when using copy_io.
  Pass exceptions back to main fiber when using copy_io.
  Attempt to kill process if exception occurs in run block.
  Chdir param is back.
  New env modification params (ruby compatible).
  New shell param.
  New method .kill.

ENV additions:
  Enumerable
  New method .clear
  New method .keys

### CODE REVIEW

File: src/env.cr [hunk 1]
```diff
@@ -37,14 +39,22 @@ module ENV
...
     environ_ptr = LibC.environ
     while environ_ptr
       environ_value = environ_ptr.value
       if environ_value
         key_value = String.new(environ_value)
-        key, value = key_value.split '=', 2
-        yield key, value
+        kv = key_value.split '=', 2
+        key = kv[0]
+        val = kv[0]? || ""
```
Comment (jhass) [inline src/env.cr]: meant to be `[1]`?

File: src/env.cr [hunk 2]
```diff
@@ -37,14 +39,22 @@ module ENV
...
+        yield({key, val})
+# one day...
+#        yield(key, val)
```
Comment (jhass) [inline src/env.cr]: Don't think we need this comment here :)

File: src/process/run.cr [hunk 3]
```diff
@@ -10,12 +10,13 @@ class Process
...
   # * `IO`: use the given IO
   alias Stdio = Nil | Bool | IO
+  alias Env = Nil | Hash(String, Nil) | Hash(String, String?) |  Hash(String, String)
 
   # Executes a process and waits for it to complete.
   #
   # By default the process is configured without input, output or error.
-  def self.run(cmd : String, args = nil, input = false : Stdio, output = false : Stdio, error = false : Stdio) : Process::Status
-    new(cmd, args, input, output, error).wait
+  def self.run(cmd : String, args = nil, env = nil : Env, unsetenv_others = false : Bool, shell = false : Bool, input = false : Stdio, output = false : Stdio, error = false : Stdio, chdir = nil : String?) : Process::Status
```
Comment (jhass) [inline src/process/run.cr]: How about calling `chdir` `working_directory` ?

Comment (jhass) [inline src/process/run.cr]: A better name for `unsetenv_others` would be `inherit_env` and flip the semantic IMO.

File: src/process/run.cr [hunk 4]
```diff
@@ -93,11 +106,21 @@ class Process
         reopen_io(fork_output || output, STDOUT, "w")
         reopen_io(fork_error || error, STDERR, "w")
 
-        # Dir.chdir(chdir) if chdir
+        ENV.clear if unsetenv_others
+        if env
+          env.not_nil!.each do |key, val|
```
Comment (jhass) [inline src/process/run.cr]: Did you test if the `.not_nil!` is still needed after the `env = Env` bug was fixed? I doubt it is.

File: src/process/run.cr [hunk 5]
```diff
@@ -110,12 +133,23 @@ class Process
     fork_error.try &.close
   end
 
-  # Waits for this process to complete.
+  def kill sig = Signal::TERM
```
Comment (jhass) [inline src/process/run.cr]: A simple doc comment would be nice, perhaps just "Send a signal to the process, see Process.kill"

**Comment by asterite** [comment]
Thanks!

But I think it's better if you send small pull requests with each addition.

For example I don't think we need the "shell" argument, because you can do `system` and "backtick", which do that. If you need to invoke the shell it's not a big deal to do `Process.run("/bin/sh", {"-c", command})`. Also, now there's a runtime exception if you pass "shell" and "args", where in the previous case you wasn't able to trigger one.

But I think the other stuff is OK (I didn't look through all of it). But since I'm not sure about one thing, we can't merge this pull request. If they are individual things we can merge each of them if we think they are OK.

File: src/process/run.cr [hunk 6]
```diff
@@ -46,12 +53,18 @@ class Process
   # To wait for it to finish, invoke `wait`.
   #
   # By default the process is configured without input, output or error.
-  def initialize(command : String, args = nil, input = false : Stdio, output = false : Stdio, error = false : Stdio)
-    argv = [command.to_unsafe]
-    args.try &.each do |arg|
-      argv << arg.to_unsafe
+  def initialize(command : String, args = nil, env = nil : Env, unsetenv_others = false : Bool, shell = false : Bool, input = false : Stdio, output = false : Stdio, error = false : Stdio, chdir = nil : String?)
+    cmd, argv = if shell
+      raise "args not allowed with shell" if args
```
Comment (jhass) [inline src/process/run.cr]: Why wouldn't arguments be allowed anyway.

File: src/env.cr (hunk 2 above)
Comment (asterite) [inline src/env.cr]: `yield key, val` is just fine, I don't understand why you changed it to a tuple

Comment (jhass) [inline src/env.cr]: For Enumerable

Comment (asterite) [inline src/env.cr]: I never understood the point of Hash being Enumerable in Ruby. If you do `select`, `reject`, etc., you get arrays, and you probably wanted a Hash as a result, so you end up doing `Hash[hash.select { ... }]`, which is long to type and slow.

I think in this case I'd prefer to have specialized Hash methods that return hashes in these cases. One exception is `each_with_index`, but we already implemented that. I don't think `map` makes sense either.

And in the case you do want to treat Hash as an enumerable of tuples, you can do `hash.each`.

Comment (sergey-kucher) [inline src/env.cr]: +1

**Comment by technorama** [comment]
* Hash[array] is old syntax.  Try array.to_h in ruby 2.1
* At least some of the hash methods in ruby do return hashes

```
{ a: 1, b: 2}.select { true } 
=> {:a=>1, :b=>2} 
```

If you would like to provide specialized methods I guess we could go that route but how would that work
with type restrictions? 
```
def f a : Enumerable # fails with Hash even though all methods are implemented
```

* key, val was changed to tuple to support Enumerable.  map() is used in keys() for clear().  
Enumerable methods have many uses on a hash, especially when combined with map

Example of creating a shell script preserving specific environment variables (this is a shortened example from a real program):
```
preserve = Set.new(%w(PATH HOME USER ...))
vars = ENV.select { |k, v| preserve.includes?(k) }.sort.map { |k, v| "#{k}=#{v.shell_escape}" }
```
Alternatively you could end it with .to_h instead of map to get a hash back.

**Comment by technorama** [comment]
@jhass most of your suggestions were just committed.  unsetenv_others -> clear_env.  chdir was left alone as your suggesting seems like too much typing.

**Comment by technorama** [comment]
@asterite system and ` don't give you any control over stdin/out/err.  Stdin **can't** be used with either as it is used for sending commands in both methods.  Both methods lack a way to provide any arguments for setting environment variables (overriding PATH or other).

Providing an array of args when you expect to pass a command to a shell is not sensible.  If the runtime error can be changed to a compile time error I'm happy to make the change.  Should I open a issue for compile time assertions?  Eiffel like contracts are on my wish list.  I'd prefer those instead.

File: src/process/run.cr [hunk 7]
```diff
@@ -46,12 +53,18 @@ class Process
   # To wait for it to finish, invoke `wait`.
   #
   # By default the process is configured without input, output or error.
-  def initialize(command : String, args = nil, input = false : Stdio, output = false : Stdio, error = false : Stdio)
-    argv = [command.to_unsafe]
-    args.try &.each do |arg|
-      argv << arg.to_unsafe
+  def initialize(command : String, args = nil, env = nil : Env, clear_env = false : Bool, shell = false : Bool, input = false : Stdio, output = false : Stdio, error = false : Stdio, chdir = nil : String?)
```
Comment (jhass) [inline src/process/run.cr]: I still find `clear_env` a bit confusing and would prefer `inherit_env`. `run(..., env: {"foo": "bar"}, clear_env: true)` seems a bit confusing, I think `run(..., env: {"foo": "bar"}, inherit_env:  false)` is much clearer.

I also still think `working_directory` would be a lot clearer, I think its usecase is really rare (I never felt the need for something like that anywhere outside bash scripts yet), so I'd much prefer the clarity for rarely used feature over that little bit more typing.

**Comment by jhass** [comment]
> Providing an array of args when you expect to pass a command to a shell is not sensible.

Any arguments to back the claim? I tend to agree that it needs unnecessary noise to the already huge signature. Why is `run("foo", shell: "/bin/zsh")` better than `run("/bin/zsh", {"-c", "foo"})`, which you need to do anyway to pass arguments to `foo` in your variant.

**Comment by technorama** [comment]
The shell argument is true or false.  /bin/sh only (just like traditional popen(), system() or many other methods built on them).

The shell argument is provided for common usage, especially for those less versed in unix who expect the traditional behavior of popen() or system().  It's merely a convenience for them and documents the intent of the programmer.

An example of what doesn't make sense:
```
run("foo | grep ...", ["something", "else"], shell: true)
```

**Comment by technorama** [comment]
@asterite Enumerable was removed from ENV.  The ENV IndexError fix still exists as I can't run the tests without it.  Is this suitable for merging?

**Comment by jhass** [comment]
Let's just provide `run_shell` or something then that doesn't even take `args`?

**Comment by waj** [comment]
I like most of this PR. I also think like @jhass that the runtime exception could be avoided just providing another overload. Maybe `Process.shell`?

**Comment by technorama** [comment]
@waj There's just a small naming issue

Process.run waits
Process.run { } waits
Process.new returns a wrapped pid where you control the waiting

Process.shell waits
Process.shell { } waits
Process.what_do_i_call it to return a wrapped @pid?

**Comment by asterite** [comment]
Hm, in the beginning `Process.run` didn't wait, just the block version did. So we might consider going back to that behaviour. The "issue" was that `Process.run` and `Process.new` did the same thing, but I think I don't mind. @waj thoughts?

**Pull request merged**